source/opt/workaround_image_operands.cpp

```diff
@@ -0,0 +1,173 @@
+                          analysis::Image img(
+                              t, img_type->dim(), 1,
+                              img_type->is_arrayed(),
+                              img_type->is_multisampled(), img_type->sampled(),
+                              img_type->format(), img_type->access_qualifier());
+                          // Make a new image type identical to the previous, but 
+                          // mark off the depth bit.
+                          for (auto& dec : img_type->decorations()) {
+                            img.AddDecoration(std::vector<uint32_t>(dec));
+                          }
+                          uint32_t new_image =
+                              type_mgr->GetTypeInstruction(&img);
+
+                          // Create a new OpTypeSampledImage with the given type. 
+                          analysis::SampledImage sampled_img(
+                              type_mgr->GetType(new_image));
+                          for (auto& dec : sampled_image_type->decorations()) {
+                            sampled_img.AddDecoration(
+                                std::vector<uint32_t>(dec));
+                          }
+
+                          uint32_t sampled_image_type =
+                              type_mgr->GetTypeInstruction(&sampled_img);
+
+                          // Create a new OpSampledImage value that uses that
+                          // type.
+                          auto new_sampled_image =
+                              op_sampled_image->Clone(context());
+                          new_sampled_image->SetResultType(sampled_image_type);
+
+                          new_sampled_image->InsertBefore(op_sampled_image);
+                          new_sampled_image->SetResultId(
+                              context()->TakeNextId());
+
+                          // Update the DRef instruction to point to the new sampled image.
+                          sampled_image_usage->SetOperand(
+                              idx, {new_sampled_image->result_id()});
+                        });
+                        // TODO: Should we remove the original sampled image?
+                        //       If all of its users have been switched (common case)?
+                        // TODO: Same with OpTypeImage/OpTypeSampledImage
+                  }
+                });
+                // As-is this gives the same behavior as glslang, which is not completely
+                // correct.
+                // We should also backtrack to find the actual textures used, and switch
+                // its type to the new depth image.
+                // If we have both? not sure what the right answer is, probably pick one
+                // or the other (depth = 1 is probably safe)
+          }
+        });
+  }
+
+  for (auto img_type : unknown_depth_images) {
+    analysis::Type* t =
+        type_mgr->GetType(type_mgr->GetId(img_type->sampled_type()));
+    ir::Instruction* inst =
```

Comment:
It seems like I am mixing the ir::Instruction and analysis::Type* which seems wrong.

Reply:
This will make the TypeManager inconsistent, but we do not invalidate it.  Create a new type, and use `context->ReplaceAllUses(...)`.  This should be okay once the def-use manager is kept up-to-date.